### DISCUSSION TITLE
Getting a ServiceClient from XrmFakedContext

### DISCUSSION THREAD

**Discussion by sandeep-chinni** [DynamicsValue/fake-xrm-easy#29] [Q&A] [answered]
Carrying on the discussion from here: https://github.com/jordimontana82/fake-xrm-easy/discussions/581

@jordimontana82 I am using this newer version, I started the discussion on the above thread as it felt more relevant. I have gone through the wiki links you shared on the previous discussion and I am no closer to solving the issue I am facing, maybe because I am a newbie in the dynamics world.

Here is what my code looks like
`        [TestMethod]
        public void TestCreateWhenStatusAssignedThenRecordCreated()
        {
            
            Mock<DefaultCrmClientFactory> clientFactory = new(configuration, null);
            IXrmFakedContext context = MiddlewareBuilder
                .New()
                .AddCrud()
                .AddFakeMessageExecutors()
                .UseCrud()
                .UseMessages()
                .SetLicense(FakeXrmEasyLicense.RPL_1_5)
                .Build();

            ServiceClient client = (ServiceClient)context.GetAsyncOrganizationService();
            clientFactory.Setup(x => x.CreateClient()).Returns(client);
            //more code/testing 
       }
`
The problem here is, our implementation needs an instance of ServiceClient(Microsoft.PowerPlatform.Dataverse.Client.ServiceClient) which implements the IOrganizationService/2. When I try to the OrganizationService from context and cast it to (ServiceClient), it obviously fails because the cast from interface to concrete object is not possible as stated here:https://stackoverflow.com/questions/48773168/using-moq-system-invalidcastexception-unable-to-cast-object-of-type-castle-p
<img width="1187" alt="Screen Shot 2022-03-20 at 1 21 28 AM" src="https://user-images.githubusercontent.com/97115087/159150804-d90289cd-4f18-4e43-8320-f2bed1604676.png">

So, how do I solve this issue? All the wikis for this new repo deal only with IOrgnaizationService, I haven't seen one with ServiceClient.

PS: Regarding the license, we are still evaluating if this is the correct tool for us. If we decide it is, we will work on the licensing part.

Thanks!

**Comment by jordimontana82** [comment, marked as answer]
Hi @sandeep-chinni ,

There's no need to use Moq since ServiceClient already implements IOrganizationService/2. You need to refactor your azure function code to either use IOrganizationService or IOrganizationService2, then pass your ServiceClient instance into those methods. Doing it that way there is no need to cast anything as ServiceClient already implements those interfaces.

WIth FakeXrmEasy you could use IOrganizationService, IOrganizationServiceAsync, or IOrganizationServiceAsync2 interfaces.

Suggest starting with specific documentation about version 3:

**A specific section for AzureFunctions**: https://dynamicsvalue.github.io/fake-xrm-easy-docs/quickstart/azure-functions/. In there you'll find a repo with sample code.

**And how to use the Async interface**:  https://dynamicsvalue.github.io/fake-xrm-easy-docs/quickstart/advanced/async/

Hope this helps.

**Reply by sandeep-chinni** [reply]
@jordimontana82 I was afraid you were going to say something like that :). Thanks though, it is what is with regards to our implementation. I will explore the possibility of refactoring our code like you suggested.

I have one question about the AzureFunctions sample code: https://github.com/DynamicsValue/dataverse-summit-2021/blob/main/src/CreateContactFn.cs#L42

I do not see a place where the CreateContact method on line 42 is getting called from. I am asking because I want to know how the IOrganizationServiceAsync2 is instantiated/injected in a called method other than the unit test.

I apologize if this question falls a bit outside of your framework/utility here.

Thanks!

**Reply by jordimontana82** [reply]
@sandeep-chinni  Unless there are some methods in ServiceClient that don't exist in any of these interfaces and that you really need for your azure function?

**Reply by jordimontana82** [reply]
We were crossing answers there at the same time :)  

In the source code of the azure function: 

1) You create an instance of ServiceClient in the main body of the Azure function (this would require a connection string, that could be retrieved from keyvault, environment variables etc
2) Replace all references to ServiceClient in your methods / classes etc (other than that first ServiceClient instance above) with one of the IOrganizationService* interfaces
3) Then pass the instance from step 1 above into whatever methods were using ServiceClient that were replaced with one of the interfaces.

In the unit test project:

- You just use the default interface that is returned after building the context as explained , for example, here: https://dynamicsvalue.github.io/fake-xrm-easy-docs/quickstart/advanced/async/
It should work.

Here's some sample code updated:

https://github.com/DynamicsValue/dataverse-summit-2021/blob/main/src/CreateContactFn.cs#L34-L45

Does that make sense?

**Reply by jordimontana82** [reply]
If you have too many methods to replace or refactor I might consider adding this:

https://github.com/DynamicsValue/fake-xrm-easy/issues/30

Although it would be best to always use interfaces, unless there are methods / properties specific to the ServiceClient instance that you really need...

**Reply by jordimontana82** [reply]
@sandeep-chinni  I pushed some changes to that sample repo that covers what I explained above. Hope this helps:

https://github.com/DynamicsValue/dataverse-summit-2021/blob/main/src/CreateContactFn.cs#L34-L45

**Reply by jordimontana82** [reply]
Any luck @sandeep-chinni  ?

**Reply by sandeep-chinni** [reply]
@jordimontana82 Thank you very much for the responses. The sample code does help. I will try and see if reworking our code is an option. I will post an update as soon as I have one. And once again, thank you very much for the prompt feedback and responses!

PS: I am sorry for the delayed response, I called it a day after posting my question.

**Reply by jordimontana82** [reply]
@sandeep-chinni  Great! Glad it was helpful. Feel free to let me know once you get a chance to try it.

Any other questions let me know.

Cheers

**Reply by jordimontana82** [reply]
Hi @sandeep-chinni , how did it go? Any update?

**Reply by KannanPooja** [reply]
@jordimontana82 I will try. Thank you very much.
